making tidb upgrade from v3.0.5(tidb-ansible deploy it) to v4.0.0(tiup upgrade) but tidb don't running well

## General Question

Before asking a question, make sure you have:

- Searched existing Stack Overflow questions.
- Googled your question.
- Searched open and closed [GitHub issues](https://github.com/pingcap/tiup/issues?utf8=%E2%9C%93&q=is%3Aissue)
- Read the documentation:
  - [TiUP Readme](https://github.com/pingcap/tiup)
  - [TiUP Doc](https://github.com/pingcap/docs)

when i use tiup tool to upgrading my tidb cluster from v3.0.5 to v4.0.0.
After tiup upgrade tikv nodes successfully, then it's going to upgrade tidb nodes.Then something goes happening.
Tiup cluster tools tell tidb 4000 don't go up after 1m.Then ,i search this tidb node for some useful information below it :
```
[2020/06/11 13:37:52.908 +08:00] [INFO] [ddl_worker.go:124] ["[ddl] DDL worker closed"] [worker="worker 22, tp add index"] ["take time"=10.626µs]
[2020/06/11 13:37:52.908 +08:00] [INFO] [ddl_worker.go:124] ["[ddl] DDL worker closed"] [worker="worker 21, tp general"] ["take time"=30.891µs]
[2020/06/11 13:37:52.908 +08:00] [INFO] [delete_range.go:123] ["[ddl] closing delRange"]
[2020/06/11 13:37:52.908 +08:00] [INFO] [session_pool.go:85] ["[ddl] closing sessionPool"]
[2020/06/11 13:37:52.908 +08:00] [INFO] [ddl.go:407] ["[ddl] DDL closed"] [ID=b1a2c40b-f658-4cdf-acd2-2770e14bb24b] ["take time"=25.269518ms]
[2020/06/11 13:37:52.908 +08:00] [INFO] [ddl.go:301] ["[ddl] stop DDL"] [ID=b1a2c40b-f658-4cdf-acd2-2770e14bb24b]
[2020/06/11 13:37:52.916 +08:00] [INFO] [manager.go:267] ["failed to campaign"] ["owner info"="[ddl] /tidb/ddl/fg/owner ownerManager b1a2c40b-f658-4cdf-acd2-2770e14bb24b"] [error="context canceled"]
[2020/06/11 13:37:52.916 +08:00] [INFO] [manager.go:248] ["break campaign loop, context is done"] ["owner info"="[ddl] /tidb/ddl/fg/owner ownerManager b1a2c40b-f658-4cdf-acd2-2770e14bb24b"]
[2020/06/11 13:37:52.955 +08:00] [INFO] [manager.go:292] ["revoke session"] ["owner info"="[ddl] /tidb/ddl/fg/owner ownerManager b1a2c40b-f658-4cdf-acd2-2770e14bb24b"] []
[2020/06/11 13:37:52.985 +08:00] [INFO] [domain.go:607] ["domain closed"] ["take time"=102.9283ms]
[2020/06/11 13:37:52.985 +08:00] [ERROR] [tidb.go:82] ["[ddl] init domain failed"] [error="strconv.ParseInt: parsing \"to_seconds(\\\"20190414\\\")\": invalid syntax"]

```
but i don't know why tidb output that.
Wish you some suggestions,thank you very much!

--- a/table/tables/partition.go
+++ b/table/tables/partition.go
@@ -163,7 +163,7 @@ type ForRangePruning struct {
 }
 
 // dataForRangePruning extracts the less than parts from 'partition p0 less than xx ... partitoin p1 less than ...'
-func dataForRangePruning(pi *model.PartitionInfo) (*ForRangePruning, error) {
+func dataForRangePruning(sctx sessionctx.Context, pi *model.PartitionInfo) (*ForRangePruning, error) {
 	var maxValue bool
 	var unsigned bool
 	lessThan := make([]int64, len(pi.Definitions))
@@ -182,7 +182,12 @@ func dataForRangePruning(pi *model.PartitionInfo) (*ForRangePruning, error) {
 				unsigned = true
 			}
 			if err != nil {
-				return nil, errors.WithStack(err)
+				val, ok := fixOldVersionPartitionInfo(sctx, pi.Definitions[i].LessThan[0])
+				if !ok {
+					logutil.BgLogger().Error("wrong partition definition", zap.String("less than", pi.Definitions[i].LessThan[0]))
+					return nil, errors.WithStack(err)
+				}
+				lessThan[i] = val
 			}
 		}
 	}
@@ -191,6 +196,17 @@ func dataForRangePruning(pi *model.PartitionInfo) (*ForRangePruning, error) {
 		MaxValue: maxValue,
 		Unsigned: unsigned,
 	}, nil
+}
+
+func fixOldVersionPartitionInfo(sctx sessionctx.Context, str string) (int64, bool) {
+	// less than value should be calculate to integer before persistent.
+	// Old version TiDB may not do it and store the raw expression.
+	tmp, err := parseSimpleExprWithNames(parser.New(), sctx, str, nil, nil)
+	ret, isNull, err := tmp.EvalInt(sctx, chunk.Row{})
+	if err == nil && !isNull {
+		return ret, true
+	}
+	return ret, false
 }
 
 // rangePartitionString returns the partition string for a range typed partition.
@@ -240,7 +256,7 @@ func generateRangePartitionExpr(ctx sessionctx.Context, pi *model.PartitionInfo,
 
 	switch len(pi.Columns) {
 	case 0:
-		tmp, err := dataForRangePruning(pi)
+		tmp, err := dataForRangePruning(ctx, pi)
 		if err != nil {
 			return nil, errors.Trace(err)
 		}

--- a/table/tables/partition.go
+++ b/table/tables/partition.go
@@ -202,11 +202,14 @@ func fixOldVersionPartitionInfo(sctx sessionctx.Context, str string) (int64, boo
 	// less than value should be calculate to integer before persistent.
 	// Old version TiDB may not do it and store the raw expression.
 	tmp, err := parseSimpleExprWithNames(parser.New(), sctx, str, nil, nil)
-	ret, isNull, err := tmp.EvalInt(sctx, chunk.Row{})
-	if err == nil && !isNull {
-		return ret, true
+	if err != nil {
+		return 0, false
 	}
-	return ret, false
+	ret, isNull, err := tmp.EvalInt(sctx, chunk.Row{})
+	if err != nil || isNull {
+		return 0, false
+	}
+	return ret, true
 }
 
 // rangePartitionString returns the partition string for a range typed partition.

